### DISCUSSION TITLE
Something like Hoogle?

### DISCUSSION THREAD

**Discussion by atifaziz** [morelinq/MoreLINQ#996] [General]
This discussion was forked out of PR #989.

@declard [wrote](https://github.com/morelinq/MoreLINQ/pull/989#issuecomment-1486546773):

> Just an idea: in Haskell ecosystem (there is `groupBy` func with the same type signature as the one this PR adds, BTW) there is a tool called Hoogle which allows to search functions across libraries by their type and it's tremendously convenient, may it make sense to add type annotations to the functions descriptions in [README.md](https://github.com/morelinq/MoreLINQ/blob/master/README.md)? Like
> 
> ```
> Segment
> [a] -> (a -> Bool) -> [[a]]
> [a] -> (a -> Int -> Bool) -> [[a]]
> [a] -> (a -> a -> Int -> Bool) -> [[a]]
> ```
> 
> The syntax is alien to C#, so there can be a better option, but it is extremely concise. I use similar pseudo-typelanguage in the extension lib on the project.
> 
> I guess I should open an issue for such discussion, don't I?

@viceroypenguin [wrote](https://github.com/morelinq/MoreLINQ/pull/989#issuecomment-1487006108):

> That information is available in the public docs: https://morelinq.github.io/3.4/ref/api/html/Overload_MoreLinq_MoreEnumerable_Segment.htm

@declard [wrote](https://github.com/morelinq/MoreLINQ/pull/989#issuecomment-1487660453):

> > That information is available in the public docs: https://morelinq.github.io/3.4/ref/api/html/Overload_MoreLinq_MoreEnumerable_Segment.htm
> 
> yes, but imagine one comes into the docs, they want to find a suitable function by its type, they press Ctrl+F and type... what?

**Comment by atifaziz** [comment]
> yes, but imagine one comes into the docs, they want to find a suitable function by its type, they press Ctrl+F and type... what?

I know [Hoogle](https://hoogle.haskell.org/) and it works really well for a Haskell developer who think in terms of types and function signatures and have built proficiency in reading those. When I was actively developing in [LiveScript](https://livescript.net/) and using [its prelude library](http://www.preludels.com/), I really liked how it concisely expressed the logical transformation as simple signatures, [replicate](http://www.preludels.com/#replicate), like [`replicate: Number → a → [a]`](http://www.preludels.com/#replicate), and you'll notice I've used those in title of some issues. I think doing something like that for MoreLINQ will serve a tiny fraction of the community. Perhaps something more useful would be group the various extensions instead of having a single flat list, like how it's done in the C# documentation section title, “[Standard Query Operators Overview (C#)](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/standard-query-operators-overview)”:

- [Sorting Data](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/sorting-data.md)
- [Set Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/set-operations.md)
- [Filtering Data](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/filtering-data.md)
- [Quantifier Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/quantifier-operations.md)
- [Projection Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/projection-operations.md)
- [Partitioning Data](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/partitioning-data.md)
- [Join Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/join-operations.md)
- [Grouping Data](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/grouping-data.md)
- [Generation Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/generation-operations.md)
- [Equality Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/equality-operations.md)
- [Element Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/element-operations.md)
- [Converting Data Types](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/converting-data-types.md)
- [Concatenation Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/concatenation-operations.md)
- [Aggregation Operations](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/aggregation-operations.md)

**Reply by viceroypenguin** [reply]
@atifaziz This is a good idea. I've implemented it in viceroypenguin/SuperLinq#359. You can observe the results [here](https://github.com/viceroypenguin/SuperLinq/blob/bf25b9d7d5e3f8e2fee6abbb7843a50c479d5bcf/README.md). Feel free to copy it back to morelinq if it's useful; some work will be required (mangling the doc links, remove async links, removing operators not implemented in morelinq).
